### DISCUSSION TITLE
Fatal interaction with WP-Optimize?

### DISCUSSION THREAD

**Discussion by craigconstantine** [Automattic/wordpress-activitypub#869] [Q&A] [answered]
It appears something ActivityPub is doing, pokes some un-handled edge case in WP-Optimize. I've also posted in WP-Optimize's support area: https://wordpress.org/support/topic/fatal-error-4728/

Encountering an error in WP-Optimize code, but toggling the ActivityPub plugin off fixes the problem. WP v6.6.1, latest WP-Optimize, ActivityPub v3.1.0

```
[20-Aug-2024 11:39:30 UTC] PHP Fatal error:  Uncaught TypeError: implode(): Argument #1 ($array) must be of type array, string given in /home/open7/public_html/wp-content/plugins/wp-optimize/minify/class-wp-optimize-minify-front-end.php:1383
Stack trace:
#0 /home/open7/public_html/wp-content/plugins/wp-optimize/minify/class-wp-optimize-minify-front-end.php(1383): implode('', NULL)
#1 /home/open7/public_html/wp-includes/class-wp-hook.php(324): WP_Optimize_Minify_Front_End->process_header_scripts('')
#2 /home/open7/public_html/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#3 /home/open7/public_html/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#4 /home/open7/public_html/wp-includes/script-loader.php(2215): do_action('wp_print_script...')
#5 /home/open7/public_html/wp-includes/class-wp-hook.php(324): wp_print_head_scripts('')
#6 /home/open7/public_html/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(Array, Array)
#7 /home/open7/public_html/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#8 /home/open7/public_html/wp-includes/general-template.php(3065): do_action('wp_head')
#9 /home/open7/public_html/wp-includes/template-canvas.php(17): wp_head()
#10 /home/open7/public_html/wp-includes/template-loader.php(106): include('/home/open7/pub...')
#11 /home/open7/public_html/wp-blog-header.php(19): require_once('/home/open7/pub...')
#12 /home/open7/public_html/index.php(17): require('/home/open7/pub...')
#13 {main}
  thrown in /home/open7/public_html/wp-content/plugins/wp-optimize/minify/class-wp-optimize-minify-front-end.php on line 1383
```

I have two separate Wordpress installations (not multi-site, fully separate) exhibiting the same error. Both sites have the ActivityPub plugin installed. The error occurs only on the page (on each site) where I Have an ActivityPub shortcode. Disabling the ActivityPub plugin enables the page to load without this error.

**Comment by pfefferle** [comment]
I have checked the minify frontend class and it seems that there is a dependency mechanism that is triggering the error, but I can't see any relations to the ActivityPub plugin.

Btw. it is currently not possible to use the caching mechanism of WP-Optimize together with the ActivityPub plugin. Here is why: https://wordpress.org/support/topic/caching-conflict-between-wordpress-post-and-other-output/

**Reply by craigconstantine** [reply]
> Btw. it is currently not possible to use the caching mechanism of WP-Optimize together with the ActivityPub plugin.

I'm aware, thanx. I'm only using WP-Optimize to optimize js, ccs, browser connection count, etc.

**Reply by kowsar89** [reply]
Hi @pfefferle, developer from WP-Optimize here. I believe we've resolved the caching issue with ActivityPub in WP-Optimize v3.3.1 by incorporating a `application/activity+json` check here: https://plugins.trac.wordpress.org/browser/wp-optimize/tags/3.5.0/cache/file-based-page-cache-functions.php#L632

**Reply by pfefferle** [reply]
@kowsar89 this is awesome, thanks a lot! can you maybe also add `application/ld+json` as alternate header? ActivityPub supports to valid `Accept` headers:

* `application/ld+json; profile="https://www.w3.org/ns/activitystreams"`
* `application/activity+json`

See: https://www.w3.org/TR/activitypub/#client-to-server-interactions

Here is the check we use in the ActivityPub plugin:

```PHP
/*
 * The other (more common) option to make an ActivityPub request
 * is to send an Accept header.
 */
if ( isset( $_SERVER['HTTP_ACCEPT'] ) ) {
	$accept = sanitize_text_field( wp_unslash( $_SERVER['HTTP_ACCEPT'] ) );

	/*
	 * $accept can be a single value, or a comma separated list of values.
	 * We want to support both scenarios,
	 * and return true when the header includes at least one of the following:
	 * - application/activity+json
	 * - application/ld+json
	 * - application/json
	 */
	if ( preg_match( '/(application\/(ld\+json|activity\+json|json))/i', $accept ) ) {
		return true;
	}
}
```

Thanks again for your support!

**Reply by kowsar89** [reply]
@pfefferle I've added this to our to-do list

**Reply by pfefferle** [reply]
Thanks a lot! I will add WP-Optimize to the support page!

**Reply by pfefferle** [reply]
@kowsar89 https://github.com/Automattic/wordpress-activitypub/wiki/Caching

**Reply by kowsar89** [reply, marked as answer]
@pfefferle **_Update_**: WP-Optimize v4.0.0 now uses `preg_match('/(application\/(ld\+json|activity\+json|json))/i', ...)`. Ref: https://plugins.trac.wordpress.org/browser/wp-optimize/tags/4.0.0/cache/file-based-page-cache-functions.php#L634
